<a href="https://colab.research.google.com/github/dhtolv/Mini-Project-AVD-1/blob/main/Mini_Project_AVD_Dhita_Olivia.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Studi Kasus**

Studi kasus ini membahas tentang memahami preferensi konsumen dan meningkatkan kualitas produk cokelat berdasarkan histori rating cokelat batangan. Data ini juga berisi tentang informasi tentang berbagai karakteristik cokelat, seperti presentasi kakao, asal biji kakao, dan rating yang diberikan pelanggan.

**BUSINESS UNDERSTANDING**

**Business Objective**

Tujuan utama dataset ini adalah memahami tren pelanggan dan pasar cokelat untuk meningkatkan kualitas produk cokelat yang sesuai dengan preferensi pelanggan. Selain itu, dengan menggunakan data dapat membuat strategi pemasaran yang efektif agar menciptakan produk yang sesuai dengan selera pasar untuk meningkatkan kepuasan pelanggan.

**Assess Situation**

Dari analisis ini dapat dinilai bahwa rating produk memiliki potensi besar untuk memberikan informasi penting tentang preferensi pelanggan terhadap cokelat batangan. Namun, beberapa masalah perlu ditangani, seperti kualiatas kakao dan asal biji kakao. Oleh karena itu, memperbaiki kualitas produk sangat penting untuk meningkatkan kepuasan pelanggan dengan memilih produk yang diminati.

**Data Mining Goals**

Menganalisis tren pelanggan dan pasar cokelat untuk memperkirakan perubahan preferensi pelanggan dan mengidentifikasi produk dengan rating sangat tinggi atau sangat rendah untuk mengetahui produk yang sesuai dengan selera pasar.

**Project Plan**

Tahapan Project Plan:
1. Menentukan tujuan
2. Menerapkan timeline
3. Memilih sumber daya
4. Menentukan hasil
6. Menganalisis resiko

Dengan tujuan meningkatkan kualitas cokelat dan kepuasan pelanggan, project plan ini bertujuan untuk menganalisis data. Project plan diharapkan dapat memberikan insight yang bermanfaat dan diterapkan bagi pelanggan dengan mengikuti prosedur dan timeline yang diatur.

**DATA PREPARATION**

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('/content/flavors_of_cacao.csv')

df

Company \n(Maker-if known) Specific Bean Origin\nor Bar Name   REF  \
0                      A. Morin                       Agua Grande  1876   
1                      A. Morin                             Kpime  1676   
2                      A. Morin                            Atsane  1676   
3                      A. Morin                             Akata  1680   
4                      A. Morin                            Quilla  1704   
...                         ...                               ...   ...   
1790                     Zotter                              Peru   647   
1791                     Zotter                             Congo   749   
1792                     Zotter                      Kerala State   749   
1793                     Zotter                      Kerala State   781   
1794                     Zotter                Brazil, Mitzi Blue   486   

      Review\nDate Cocoa\nPercent Company\nLocation  Rating Bean\nType  \
0             2016            63%            France    3.75              
1             2015            70%            France    2.75              
2             2015            70%            France    3.00              
3             2015            70%            France    3.50              
4             2015            70%            France    3.50              
...            ...            ...               ...     ...        ...   
1790          2011            70%           Austria    3.75              
1791          2011            65%           Austria    3.00  Forastero   
1792          2011            65%           Austria    3.50  Forastero   
1793          2011            62%           Austria    3.25              
1794          2010            65%           Austria    3.00              

     Broad Bean\nOrigin  
0              Sao Tome  
1                  Togo  
2                  Togo  
3                  Togo  
4                  Peru  
...                 ...  
1790               Peru  
1791              Congo  
1792              India  
1793              India  
1794             Brazil  

[1795 rows x 9 columns]

**Struktur Data**

In [42]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1795 entries, 0 to 1794
Data columns (total 8 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Company 
(Maker-if known)         1795 non-null   object 
 1   Specific Bean Origin
or Bar Name  1795 non-null   object 
 2   REF                               1795 non-null   object 
 3   Cocoa
Percent                     1795 non-null   object 
 4   Company
Location                  1795 non-null   object 
 5   Rating                            1795 non-null   float64
 6   Bean
Type                         1795 non-null   object 
 7   Broad Bean
Origin                 1794 non-null   object 
dtypes: float64(1), object(7)
memory usage: 112.3+ KB


**Statistik Deskriptif**

In [43]:
df.describe()

Rating
count  1795.000000
mean      3.190146
std       0.459812
min       1.937500
25%       2.875000
50%       3.250000
75%       3.500000
max       4.437500

In [44]:
df.corr(numeric_only=True)

Rating
Rating     1.0

**Missing Values**

In [3]:
(df.isna().sum()/len(df))*100

Company \n(Maker-if known)           0.00000
Specific Bean Origin\nor Bar Name    0.00000
REF                                  0.00000
Review\nDate                         0.00000
Cocoa\nPercent                       0.00000
Company\nLocation                    0.00000
Rating                               0.00000
Bean\nType                           0.05571
Broad Bean\nOrigin                   0.05571
dtype: float64

Melakukan pengecekan nilai kosong pada Bean\nType

In [4]:
df[df['Bean\nType'].isnull()]

Company \n(Maker-if known)        Specific Bean Origin\nor Bar Name  REF  \
1544                       Soma  Three Amigos(Chuao, Wild Bolivia, D.R.)  676   

      Review\nDate Cocoa\nPercent Company\nLocation  Rating Bean\nType  \
1544          2011            70%            Canada     4.0        NaN   

      Broad Bean\nOrigin  
1544  Ven, Bolivia, D.R.

Pada kolom Bean\nType akan diimputasi karna data yang hilang berada di angka 	0.05571

In [5]:
df['Bean\nType'].dropna().describe()

count     1794
unique      41
top           
freq       887
Name: Bean\nType, dtype: object

Berdasarkan analisis statistik, mean tidak terdapat pada kolom Bean\nType, maka dari itu kita akan menggunakan 'frequensi' sebagai gantinya

In [9]:
df['Bean\nType'] = df['Bean\nType'].fillna(df['Bean\nType'].mode()[0])

df

Company \n(Maker-if known) Specific Bean Origin\nor Bar Name   REF  \
0                      A. Morin                       Agua Grande  1876   
1                      A. Morin                             Kpime  1676   
2                      A. Morin                            Atsane  1676   
3                      A. Morin                             Akata  1680   
4                      A. Morin                            Quilla  1704   
...                         ...                               ...   ...   
1790                     Zotter                              Peru   647   
1791                     Zotter                             Congo   749   
1792                     Zotter                      Kerala State   749   
1793                     Zotter                      Kerala State   781   
1794                     Zotter                Brazil, Mitzi Blue   486   

      Review\nDate Cocoa\nPercent Company\nLocation  Rating Bean\nType  \
0             2016            63%            France    3.75              
1             2015            70%            France    2.75              
2             2015            70%            France    3.00              
3             2015            70%            France    3.50              
4             2015            70%            France    3.50              
...            ...            ...               ...     ...        ...   
1790          2011            70%           Austria    3.75              
1791          2011            65%           Austria    3.00  Forastero   
1792          2011            65%           Austria    3.50  Forastero   
1793          2011            62%           Austria    3.25              
1794          2010            65%           Austria    3.00              

     Broad Bean\nOrigin  
0              Sao Tome  
1                  Togo  
2                  Togo  
3                  Togo  
4                  Peru  
...                 ...  
1790               Peru  
1791              Congo  
1792              India  
1793              India  
1794             Brazil  

[1795 rows x 9 columns]

**Duplicated Values**

In [10]:
df[df.duplicated()]

Empty DataFrame
Columns: [Company 
(Maker-if known), Specific Bean Origin
or Bar Name, REF, Review
Date, Cocoa
Percent, Company
Location, Rating, Bean
Type, Broad Bean
Origin]
Index: []

karena tidak ada data atau kolom yang terduplikasi, maka tidak perlu penanganan

**Outliers**

In [11]:
results = []

cols = df.select_dtypes(include=['float64', 'int64'])

for col in cols:
  q1 = df[col].quantile(0.25)
  q3 = df[col].quantile(0.75)
  iqr = q3 - q1
  lower_bound = q1 - 1.5*iqr
  upper_bound = q3 + 1.5*iqr
  outliers = df[(df[col] < lower_bound) | (df[col] > upper_bound)]
  percent_outliers = (len(outliers)/len(df))*100
  results.append({'Kolom': col, 'Persentase Outliers': percent_outliers})

# Dataframe dari list hasil
results_df = pd.DataFrame(results)
results_df.set_index('Kolom', inplace=True)
results_df = results_df.rename_axis(None, axis=0).rename_axis('Kolom', axis=1)

# Tampilkan dataframe
display(results_df)

Kolom         Persentase Outliers
REF                      0.000000
Review\nDate             0.000000
Rating                   1.058496

karena nilai Rating kurang dari 10%, maka diperlukan penanganan.

melakukan imputasi

In [12]:
columns_to_impute = ["Rating"]

for col in columns_to_impute:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    # Menggunakan .loc[] agar tidak muncul SettingWithCopyWarning
    df.loc[:, col] = df[col].clip(lower=lower_bound, upper=upper_bound)

pengecekan ulang

In [13]:
results = []

cols = df.select_dtypes(include=['float64', 'int64'])

for col in cols:
  q1 = df[col].quantile(0.25)
  q3 = df[col].quantile(0.75)
  iqr = q3 - q1
  lower_bound = q1 - 1.5*iqr
  upper_bound = q3 + 1.5*iqr
  outliers = df[(df[col] < lower_bound) | (df[col] > upper_bound)]
  percent_outliers = (len(outliers)/len(df))*100
  results.append({'Kolom': col, 'Persentase Outliers': percent_outliers})

# Dataframe dari list hasil
results_df = pd.DataFrame(results)
results_df.set_index('Kolom', inplace=True)
results_df = results_df.rename_axis(None, axis=0).rename_axis('Kolom', axis=1)

# Tampilkan dataframe
display(results_df)

Kolom         Persentase Outliers
REF                           0.0
Review\nDate                  0.0
Rating                        0.0

In [20]:
df

Company \n(Maker-if known) Specific Bean Origin\nor Bar Name   REF  \
0                      A. Morin                       Agua Grande  1876   
1                      A. Morin                             Kpime  1676   
2                      A. Morin                            Atsane  1676   
3                      A. Morin                             Akata  1680   
4                      A. Morin                            Quilla  1704   
...                         ...                               ...   ...   
1790                     Zotter                              Peru   647   
1791                     Zotter                             Congo   749   
1792                     Zotter                      Kerala State   749   
1793                     Zotter                      Kerala State   781   
1794                     Zotter                Brazil, Mitzi Blue   486   

                      Review\nDate Cocoa\nPercent Company\nLocation  Rating  \
0    1970-01-01 00:00:00.000002016            63%            France    3.75   
1    1970-01-01 00:00:00.000002015            70%            France    2.75   
2    1970-01-01 00:00:00.000002015            70%            France    3.00   
3    1970-01-01 00:00:00.000002015            70%            France    3.50   
4    1970-01-01 00:00:00.000002015            70%            France    3.50   
...                            ...            ...               ...     ...   
1790 1970-01-01 00:00:00.000002011            70%           Austria    3.75   
1791 1970-01-01 00:00:00.000002011            65%           Austria    3.00   
1792 1970-01-01 00:00:00.000002011            65%           Austria    3.50   
1793 1970-01-01 00:00:00.000002011            62%           Austria    3.25   
1794 1970-01-01 00:00:00.000002010            65%           Austria    3.00   

     Bean\nType Broad Bean\nOrigin  
0                         SAO TOME  
1                             TOGO  
2                             TOGO  
3                             TOGO  
4                             PERU  
...         ...                ...  
1790                          PERU  
1791  Forastero              CONGO  
1792  Forastero              INDIA  
1793                         INDIA  
1794                        BRAZIL  

[1795 rows x 9 columns]

**Inconsistent Data**

In [29]:
df["Broad Bean\nOrigin"] = df["Broad Bean\nOrigin"].str.upper()

df

Company \n(Maker-if known) Specific Bean Origin\nor Bar Name   REF  \
0                      A. Morin                       Agua Grande  1876   
1                      A. Morin                             Kpime  1676   
2                      A. Morin                            Atsane  1676   
3                      A. Morin                             Akata  1680   
4                      A. Morin                            Quilla  1704   
...                         ...                               ...   ...   
1790                     Zotter                              Peru   647   
1791                     Zotter                             Congo   749   
1792                     Zotter                      Kerala State   749   
1793                     Zotter                      Kerala State   781   
1794                     Zotter                Brazil, Mitzi Blue   486   

                       Review\nDate Cocoa\nPercent Company\nLocation  Rating  \
0     1970-01-01 00:00:00.000002016            63%            France    3.75   
1     1970-01-01 00:00:00.000002015            70%            France    2.75   
2     1970-01-01 00:00:00.000002015            70%            France    3.00   
3     1970-01-01 00:00:00.000002015            70%            France    3.50   
4     1970-01-01 00:00:00.000002015            70%            France    3.50   
...                             ...            ...               ...     ...   
1790  1970-01-01 00:00:00.000002011            70%           Austria    3.75   
1791  1970-01-01 00:00:00.000002011            65%           Austria    3.00   
1792  1970-01-01 00:00:00.000002011            65%           Austria    3.50   
1793  1970-01-01 00:00:00.000002011            62%           Austria    3.25   
1794  1970-01-01 00:00:00.000002010            65%           Austria    3.00   

     Bean\nType Broad Bean\nOrigin  
0                         SAO TOME  
1                             TOGO  
2                             TOGO  
3                             TOGO  
4                             PERU  
...         ...                ...  
1790                          PERU  
1791  Forastero              CONGO  
1792  Forastero              INDIA  
1793                         INDIA  
1794                        BRAZIL  

[1795 rows x 9 columns]

**Construct Data**

In [32]:
df["high_flavor"] = pd.to_numeric(df["Review\nDate"], errors='coerce') / (df["Rating"] + 1)

In [33]:
df

Company \n(Maker-if known) Specific Bean Origin\nor Bar Name   REF  \
0                      A. Morin                       Agua Grande  1876   
1                      A. Morin                             Kpime  1676   
2                      A. Morin                            Atsane  1676   
3                      A. Morin                             Akata  1680   
4                      A. Morin                            Quilla  1704   
...                         ...                               ...   ...   
1790                     Zotter                              Peru   647   
1791                     Zotter                             Congo   749   
1792                     Zotter                      Kerala State   749   
1793                     Zotter                      Kerala State   781   
1794                     Zotter                Brazil, Mitzi Blue   486   

                       Review\nDate Cocoa\nPercent Company\nLocation  Rating  \
0     1970-01-01 00:00:00.000002016            63%            France    3.75   
1     1970-01-01 00:00:00.000002015            70%            France    2.75   
2     1970-01-01 00:00:00.000002015            70%            France    3.00   
3     1970-01-01 00:00:00.000002015            70%            France    3.50   
4     1970-01-01 00:00:00.000002015            70%            France    3.50   
...                             ...            ...               ...     ...   
1790  1970-01-01 00:00:00.000002011            70%           Austria    3.75   
1791  1970-01-01 00:00:00.000002011            65%           Austria    3.00   
1792  1970-01-01 00:00:00.000002011            65%           Austria    3.50   
1793  1970-01-01 00:00:00.000002011            62%           Austria    3.25   
1794  1970-01-01 00:00:00.000002010            65%           Austria    3.00   

     Bean\nType Broad Bean\nOrigin  high_flavor  
0                         SAO TOME          NaN  
1                             TOGO          NaN  
2                             TOGO          NaN  
3                             TOGO          NaN  
4                             PERU          NaN  
...         ...                ...          ...  
1790                          PERU          NaN  
1791  Forastero              CONGO          NaN  
1792  Forastero              INDIA          NaN  
1793                         INDIA          NaN  
1794                        BRAZIL          NaN  

[1795 rows x 10 columns]

**Data Reduction**

In [40]:
df = df.drop(columns=['high_flavor'])

In [41]:
df

Company \n(Maker-if known) Specific Bean Origin\nor Bar Name   REF  \
0                      A. Morin                       Agua Grande  1876   
1                      A. Morin                             Kpime  1676   
2                      A. Morin                            Atsane  1676   
3                      A. Morin                             Akata  1680   
4                      A. Morin                            Quilla  1704   
...                         ...                               ...   ...   
1790                     Zotter                              Peru   647   
1791                     Zotter                             Congo   749   
1792                     Zotter                      Kerala State   749   
1793                     Zotter                      Kerala State   781   
1794                     Zotter                Brazil, Mitzi Blue   486   

     Cocoa\nPercent Company\nLocation  Rating Bean\nType Broad Bean\nOrigin  
0               63%            France    3.75                      SAO TOME  
1               70%            France    2.75                          TOGO  
2               70%            France    3.00                          TOGO  
3               70%            France    3.50                          TOGO  
4               70%            France    3.50                          PERU  
...             ...               ...     ...        ...                ...  
1790            70%           Austria    3.75                          PERU  
1791            65%           Austria    3.00  Forastero              CONGO  
1792            65%           Austria    3.50  Forastero              INDIA  
1793            62%           Austria    3.25                         INDIA  
1794            65%           Austria    3.00                        BRAZIL  

[1795 rows x 8 columns]